### ISSUE TITLE
4-key shortcuts doesn't work

### ISSUE THREAD

**Issue by WebShapedBiz** [elementary/settings-keyboard#197]
Hi guys.

In LibreOffice Draw, the Shape -> Ungroup shortcut Shift+Ctrl+Alt+G doesn't work. At first, I thought it was something to do with my computer and/or LibreOffice but then I found a post on EOS StackExchange [here](https://goo.gl/bKTBwQ) with the same issue with IntellIJ.
All 3-keys shortcuts that I'm using are working fine.

**Comment by santileortiz** [comment]
In the keyboard plug, is "Switch Layout" set to anything other than "Disabled"?

**Comment by WebShapedBiz** [comment]
It is set by default to "Alt + Shift", I've never touched it.

**Comment by santileortiz** [comment]
Do you need it? I think disabling this should fix the issue.

**Comment by WebShapedBiz** [comment]
You're right, but I wouldn't call that a fix. I believe that this should be addressed properly.

Thank you anyhow.

**Comment by peteruithoven** [comment]
So having set the "Switch Layout" to Something like <kbd>Alt</kbd> + <kbd>Shift</kbd> can interfere with 4 key shortcuts? How does that work?

**Comment by santileortiz** [comment]
Yeah this happens because layout switching keybindings are consumed by the window manager (note I say the WM, as this does not happen in Gala's codebase, but some of its dependencies I don't remember if it's Clutter or Mutter), which effectively intercepts them before they even get to 3rd party apps.

Sadly the implementation of these keybindings is not release based and that's an issue that has been open for a long time in several projects, see https://bugs.launchpad.net/ubuntu/+source/gnome-settings-daemon/+bug/1245473 for example.

The reason why these keybindings are not "normal" is that the API available to the window manager assumes keybindings are of the kind "modifiers + unicode-character", layout switching keybindings only contain modifiers. This worked fine for a long time because X11 did all the work of compiling multiple keyboard layouts into one with a feature called groups, correctly consuming keybindings and switching layouts without the WM even knowing about these keybindings. The problem is X11 doesn't know about advanced input methods (layouts) for CYK languages.

**Comment by peteruithoven** [comment]
Is there any way we can request layout changes manually? Implement the keyboard shortcuts ourselves?

**Comment by santileortiz** [comment]
The Right Thing™ to do would be to fix the issue in Mutter, which I'm not even sure it's as simple as making modifier only keybindings release based. What happens if the user presses <kbd>Shift</kbd>+<kbd>Crtl</kbd>+<kbd>Alt</kbd>+<kbd>G</kbd> then releases <kbd>G</kbd> and <kbd>Ctrl</kbd>?, releasing <kbd>Alt</kbd> or <kbd>Shift</kbd> would cause an unintended layout switch, how can Mutter detect that?.

But anyway, assuming we get the ideal behavior in Mutter, then we need to make Gala use the latest version of Mutter, which I think is non trivial.

Another alternative is to reimplement keybinding handling in Gala's codebase, but I have no idea of how big of an undertaking this is, and it would be "duplicating functionality" of Mutter, which is generally frowned upon.

All this being said, I think a good solution, at least for this specific issue, would be to set the default keybinding to "Disabled", so we don't break keybindings for people that don't even know layout switching exists.

**Comment by santileortiz** [comment]
> Is there any way we can request layout changes manually?

Note that Gala does request layout changes to Mutter (request that is then forwarded to Clutter, and then translated into X11's API, which causes https://github.com/elementary/gala/issues/220). The problem here is the code inside Mutter that detects when a modifier-only keybinding is triggered.

**Comment by lenemter** [comment]
Closing because I cannot reproduce the issue in OS 8
